### DISCUSSION TITLE
My very first experience and propositions

### DISCUSSION THREAD

**Discussion by aeore** [Part-DB/Part-DB-server#643] [Ideas]
Hello

I have started using this project several days ago and generally I like it, but there are several problems making this project confusingly useless in some cases.
Before this project I have used confluence for such purpose, but in general confluence is just a wiki system and absolutely not designed for parts management.

So here is my list with lacks of functionality I'm observing:
1. Sorting options for category
   Will be really great if user could choose sorting options for the each selectable category. I generally like the default sorting behavior, but sometimes it is confusing and makes some categories really useless for visual assessment.
   For example, my inductor list looks like this:
```
    1 nH	
    1 uH
    1.5 uH	
    1.8 nH	
    1.8 uH	
    10 nH	
    100 nH	
    12 nH	
    15 nH	
    150 nH	
    18 nH	
    2.2 uH	
    2.7 uH	
    22 nH
    ...
```
I think with resistance and capacitance situation will be the same. I have tried enabling DATABASE_EMULATE_NATURAL_SORT, but there was no difference.
This issue is very annoying and really needs attention.

2. Separations in the table
   This feature is not so important, but will be great if system automatically separate details in the table based on the value degree, like:
```
   1 nH
   1.5 nH
   1.8 nH
   ...
   ------- <- separator
   10 nH
   15 nH
   18 nH
   ...
   ------- <- separator
   100 nH
   110 nH
   150 nH
   ...
   etc.
```
   This feature is related to the previous one and could be optional in case of special sorting is enabled.

3. Text colors in the table
   Sometimes it is good to use font foreground/background colors to mark some parts like leds. And this works in the part description, but in the table view all colors are changing to the yellow.
   Also will be good to add possibilities to write not only the text, but something like customizable `statuses`. You already have something similar when printing true/false values in the "System/Server Info" page. Also this feature will be useful with next features.

4. Custom columns
   Will be good to have possibility to add some additional custom records/columns support for some categories and allow view them in the table.

5. Table column filters for each category
   Currently once chosen column filter is used for all categories. Will be good if system store different filters for different categories
   This feature is working with previous.

6. Bulk add
   Adding a lot of similar elements is tedious. Will be good having some bulk add mechanism

7. Word wrap in the table looks not good
   When description is big, system wrap words to new line and table stay looking not very good after. Need to do something with this.

8. Custom footer/reference data region for the each category view
   Sometimes it is good to have some general description for the category with some basic reference information which includes text, pictures and, maybe, attached documents. And it will be good to see it under the table in the table view.
   
**The first feature is really MOST IMPORTANT.**

**Comment by jbtronics** [comment]
> I think with resistance and capacitance situation will be the same. I have tried enabling DATABASE_EMULATE_NATURAL_SORT, but there was no difference.
> This issue is very annoying and really needs attention.

What database are you using (the exact version)? Have you cleared the application cache after setting the env var?

**Reply by aeore** [reply]
MySQL 8.4.0

>  Have you cleared the application cache after setting the env var?

I'm using docker, so I have just re-created the container

This is docker command:
```
sudo docker run -d --name=partdb --restart unless-stopped \
-v /volume2/docker/partdb/uploads:/var/www/html/uploads \
-v /volume2/docker/partdb/public_media:/var/www/html/public/media \
-v /volume2/docker/partdb/db:/var/www/html/var/db \
-e DATABASE_URL=mysql://partdb:partdb@database:3306/partdb \
-e DATABASE_EMULATE_NATURAL_SORT=1 \
-e APP_ENV=docker \
-e DEFAULT_LANG=en \
-e DEFAULT_TIMEZONE=Europe/Berlin \
-e BASE_CURRENCY=EUR \
-e INSTANCE_NAME=Part-DB \
-e ALLOW_ATTACHMENT_DOWNLOADS=0 \
-e USE_GRAVATAR=1 \
-e DEFAULT_URI=<url is here> \
-e TABLE_PARTS_DEFAULT_COLUMNS=name,description,category,footprint,manufacturer,storage_location,amount,attachments \
-p 8095:80 \
--link mysql8:database \
jbtronics/part-db1:latest
```

**Reply by aeore** [reply]
> > I think with resistance and capacitance situation will be the same. I have tried enabling DATABASE_EMULATE_NATURAL_SORT, but there was no difference.
> > This issue is very annoying and really needs attention.
> 
> What database are you using (the exact version)? Have you cleared the application cache after setting the env var?

do you have any ideas?

**Reply by jbtronics** [reply]
Are you sure that you are really using Part-DB 1.13.0 ?

**Reply by aeore** [reply]
> Are you sure that you are really using Part-DB 1.13.0 ?

I'm using Part-DB 1.13.1
I have configured and started it without DATABASE_EMULATE_NATURAL_SORT, but later have added it when found sort issues. But I don't see any reaction

**Reply by aeore** [reply]
@jbtronics, If it is difficult to introduce stable natural sorting, may be will be good to implement manual sorting as an option? Will be good if this could be choosen for the category and manual drag-n-drop sorting would become avaialble.

**Reply by jbtronics** [reply]
@aeore You should try to update to Part-DB 1.13.2 and upgrade your database schema afterwards (`bin/console doctrine:migrations:migrate`). There was some issue with the natural sort emulation function for MySQL, which is now fixed and might make it work for you.

If it still does not work, you should look into the server info page in the "databse tab", what is stated as "natural sort method".

**Reply by aeore** [reply]
Hi. Did it, but it says natural sort is not enabled:
![изображение](https://github.com/user-attachments/assets/268be47f-a522-4583-9c91-38b7d6dd4726)

Also I have tried to check if environment variable is available - and it is avaialble:
`sudo docker exec -it partdb env <enter>
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
...
DATABASE_EMULATE_NATURAL_SORT=1
...`
